contracts/settling_game/L06_Combat.cairo

```diff
@@ -38,36 +73,154 @@ end
+        resource_len, resource_ids, resource_values)
+    let (token_ids : Uint256*) = alloc()
+    let (token_values : Uint256*) = alloc()
+    convert_cost_resources_to_unique_tokens(d_len, d, token_ids, token_values)
+
+    # pay for the squad
+    IERC1155.burn_batch(resource_address, caller, d_len, token_ids, d_len, token_values)
+
+    # assemble the squad
+    # TODO:
+    IS06_Combat.assemble_squad_from_troops_in_realm(combat_state_address)
+
+    # TODO: emit an event?
+
+    return ()
+end
+
+@external
+func initiate_combat{range_check_ptr, syscall_ptr : felt*, pedersen_ptr : HashBuiltin*}(
+        attacking_realm_id : Uint256, defending_realm_id : Uint256, attack_type : felt) -> (
+        combat_outcome : felt):
+    alloc_locals
+
+    # TODO: uncomment this when the creation of the combat contracts in the test suite is proper
+    # with_attr error_message("caller is not attacking realm owner"):
+    #     let (caller) = get_caller_address()
+    #     let (controller_address_) = controller_address.read()
+    #     let (realms_address) = IModuleController.get_realms_address(contract_address=controller_address_)
+    #     let (owner) = realms_IERC721.ownerOf(contract_address=realms_address, token_id=attacking_realm_id)
+    #     assert caller = owner
+    # end
+
+    with_attr error_message("cannot initiate combat"):
+        let (can_attack) = Realm_can_be_attacked(attacking_realm_id, defending_realm_id)
+        assert can_attack = TRUE
+    end
+
+    let (controller) = controller_address.read()
+    let (combat_state_address) = IModuleController.get_module_address(
+        controller, module_id=ModuleIds.S06_Combat)
+    let (attacking_realm_data : RealmCombatData) = IS06_Combat.get_realm_combat_data(
+        controller, attacking_realm_id)
+    let (defending_realm_data : RealmCombatData) = IS06_Combat.get_realm_combat_data(
+        controller, defending_realm_id)
+
+    let (attacker : Squad) = unpack_squad(attacking_realm_data.attacking_squad)
+    let (defender : Squad) = unpack_squad(defending_realm_data.defending_squad)
+
+    let (attacker_end, defender_end, combat_outcome) = run_combat_loop(
+        attacker, defender, attack_type)
+
+    let (new_attacker : PackedSquad) = pack_squad(attacker_end)
+    let (new_defender : PackedSquad) = pack_squad(defender_end)
+
+    let new_attacking_realm_data = RealmCombatData(
+        attacking_squad=new_attacker,
+        defending_squad=attacking_realm_data.defending_squad,
+        last_attacked_at=attacking_realm_data.last_attacked_at)
+    IS06_Combat.set_realm_combat_data(controller, attacking_realm_id, new_attacking_realm_data)
+
+    let (now) = get_block_timestamp()
+    let new_defending_realm_data = RealmCombatData(
+        attacking_squad=defending_realm_data.attacking_squad,
+        defending_squad=new_defender,
+        last_attacked_at=now)
+    IS06_Combat.set_realm_combat_data(controller, defending_realm_id, new_defending_realm_data)
+
```

Comment:
This is where we call the Resources contract on successful raid and pillage 25% of the vault